tsdb/tsm1/cache_entry.go

```diff
@@ -117,6 +124,35 @@ func (e *entry) size() int {
 	return sz
 }
 
+// Contains returns true if values exist for the time interval [min, max]
+// inclusive. The entry is deduplicated prior to calling contains.
+func (e *entry) Contains(min, max int64) bool {
+	e.deduplicate()
+	e.mu.RLock()
+	contains := e.values.Contains(min, max)
+	e.mu.RUnlock()
+	return contains
+}
+
+// MaxTime returns the maximum timestamp for values in this entry.
+// The entry is deduplicated prior to calling contains.
+// If entry is empty, MaxTime will return InvalidMinNanoTime.
+func (e *entry) MaxTime() int64 {
+	e.deduplicate()
```

Comment:
It's not clear to me that you need to de-duplicate values in the cache to determine the maximum timestamp.

Reply:
This because `entry#values` doesn't guarantee sorted timestamps, so `MaxTime` would be incorrect – that is, we would be unable to use `values[len(values)-1].MaxTime()`.

Calling `Cache#Values` also deduplicates before returning the data (twice, when merging with an ongoing snapshot):

https://github.com/influxdata/influxdb/blob/a2a90ce87189c0738bddb5c9f27d71d59e060b8a/tsdb/tsm1/cache.go#L348

Upon further review, we should not put pressure on the cache to deduplicate every time this is called. I think it will be better to simply scan the values array for the `MaxTime`, being an `O(N)` operation.

Reply:
Ah I didn't realise that `deduplicate` was the only time we did a lazy sort on the cache. Yeah so we could go `O(N)` (we do try and keep the cache relatively small).

Perhaps add a note that we may need to come back to this w.r.t perf? I feel like running through the entries will be better value than sorting them all the time..

Reply:
Overall this sucks compared to having the measurements+field types cache that we have in 1.x.

Reply:
That cache didn't help us with time ranges though did it?

Reply:
No, that is a true statement. Adding that dimension will aways make this hairy.